### ISSUE TITLE
How do we debug lost packets in the ingress-sbox container?

### ISSUE THREAD

**Issue by sbrattla** [moby/moby#36032]
To my knowledge, all traffic which (1) comes in to a Docker host and (2) is sent to a VIP address representing any given service in a Docker swarm goes through the ingress-sbox container. This is a bit of a black box to me, but from what I've seen and read this is where a lot of the Docker "magic" happens.

I'd like to ask if there are any known performance limitations for the ingress-sbox container? I'm asking because we're seeing a lot of dropped packages between the docker_gwbridge and containers (tasks) behind a VIP address.

We've got a Docker Swarm where a few nginx'es and a single memcached container runs, and a fair amount of data travels between these (we captured 1.4GB data in 20 seconds today, so about 560Mbit which should be OK). All networking happens within the same host, as the Docker Swarm lives on that single swarm (no, not recommended, but it's how it is).

We see that the nginx containers does a lot of retransmitting of packages when we capture traffic on the docker_gwbridge (using Wireshark). We can't see any retransmits when we capture traffic in memcached. The more traffic, the more retransmits. The problem will thus escalate with load.

We've tuned /etc/sysctl.conf to the best of our knowledge. Many of the sysctl parameters we've tuned isn't namespaceified, so we can't set them per container (see #35993). Much of what we've done is taken from an [Oracle paper on tuning 10Gb networks.](http://www.oracle.com/technetwork/server-storage/vm/ovm3-10gbe-perf-1900032.pdf).

This is a VM running on Hyper-V. The server has 64 cores and 126GB RAM. These resources are nowhere near being used. 

Our main problem here is that we can see packages disappearing between the docker_gwbridge and memcached, so most likely inside the ingress-sbox. However, as mentioned, that's a black box. How can we even troubleshoot what's going inside it?

dmesg reports no errors.

**Comment by ddebroy** [comment]
@sbrattla do you have all the containers running within a single Linux VM inside the Hyper-V host? If so, you can try entering into the ingress_sbox sandbox in the Linux VM and start with checking out the stats and packet dumps there:

0. Enter the ingress_sbox from the Linux VM:  ```sudo nsenter --net=/run/docker/netns/ingress_sbox```

1. Dump the iptables stats for mangle and nat tables: ```iptables -vL -t mangle``` and ```iptables -vL -t nat``` . The mangle stats should show the packets being marked for load balancing based on destination port.

2. Dump ipvs stats: ```ipvsadm -L -n --stats``` which will give you an idea of the load balancing being done by ipvs using the packet marks based on iptables above.

3. Capture packets on `eth0` and `eth1` within ingress_sbox using tcpdump to see what's entering ingress_sbox and exiting it. The `eth0` dumps on 10.x is probably the most interesting.

**Comment by sbrattla** [comment]
@ddebroy thanks for your feedback. Yes, all containers run within a single Linux VM inside the Hyper-V Host.

Following your advise, we've inspected traffic running on the ingress-sbox interfaces, but we can not see the retransmissions there. So, as far as our understanding goes, this must mean that packets are being dropped/lost somewhere between the docker_gwbridge and the incoming interface on the ingress-sbox.

1. the DNAT'ing to 172.18.0.2 is taking place (as traffic indeed appears on the docker_gwbridge)
2. we can see retransmissions when we capture traffic on the docker_gwbridge
3. we can **not** see retransmissions when we capture traffic on the incoming ingress-sbox interface. 

This makes us believe that packets must be lost on the way from nginx to memcached, and specifically somewhere between the docker_gwbridge and ingress-sbox. What could explain this? 

_Could anything inside the ingress-sbox (which happens before Wireshark captures traffic) explain packet loss?_  

As a last point, to make this even more interesting, only nginx -> memcached packets are lost. If we run a few memcached connection/set/get tests from other containers or the host, all is well.

**Comment by ddebroy** [comment]
_Could anything inside the ingress-sbox (which happens before Wireshark captures traffic) explain packet loss?_

To narrow this down, how about analyzing the packet capture at the veth interface that connects docker_gwbridge to ingress-sbox?

**Comment by sbrattla** [comment]
@ddebroy we did capture traffic on the (incoming) veth interface in the ingress-sbox (the 172.18.0.2). That capture did not show any retransmissions. I don't know at what point Wireshark captures incoming traffic, but it seems that packets are dropped / lost before that. So we see the retransmissions on the docker_gwbridge, but not inside the ingress sbox.

We tried to let nginx talk to memcached using the tasks.memcached hostname today, and that works well. No dropped packets. But then again, that traffic does not touch the docker_gwbridge or ingress sbox.

MTU sizes are the same (except for the 10.X) interface on the ingress sbox which is 1450 and not 1500...but traffic seems to be dropped before that).

I need to check upagain on the dropped counters of ifconfig, but I believe I checked them and they were at 0.
